### PR TITLE
fix: calling reply.callNotFound uncaught exception

### LINKED ISSUE #3660: Calling reply.callNotFound inside frameworkErrors can result in an uncaught exception
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

3.27.0

### Plugin version

4.5.0

### Node.js version

17.4.0

### Operating system

Linux

### Operating system version (i.e. 20.04, 11.3, 10)

20.04

### Description

Following the example below it'll instantly log an error:
```
TypeError: Cannot read properties of undefined (reading 'preHandler')
    at notFound (node_modules/fastify/lib/reply.js:707:21)
    at Reply.callNotFound (node_modules/fastify/lib/reply.js:309:3)
    at frameworkErrors (test.js:14:17)
    at onBadUrl (node_modules/fastify/fastify.js:613:14)
    at Object.handler (node_modules/find-my-way/index.js:621:33)
    at Router.lookup (node_modules/find-my-way/index.js:378:14)
    at node_modules/light-my-request/index.js:87:41
    at Request.prepare (node_modules/light-my-request/lib/request.js:132:12)
    at makeRequest (node_modules/light-my-request/index.js:87:14)
    at node_modules/light-my-request/index.js:113:7
```

If an url with an invalid uri is opened in the browser then an uncaught exception will happen.

### Steps to Reproduce

Run this example:

```
const fastifyStatic = require('fastify-static');

(async () => {
	process.on(
		'uncaughtException',
		(err) => {
			console.log('uncaughtException', err);
		}
	);
	
	const fastify = require('fastify')({
		logger: true,
		frameworkErrors: (error, req, reply) => {
			return reply.callNotFound();
		}
	});
	
	fastify.register(
		fastifyStatic,
		{
			root: '/www' // doesn't need to exit
		}
	);

	await fastify.listen(8080);
	
	await fastify.inject({
		method: 'GET',
		url: '/%'
	});
})()
	.catch((err) => {
		console.error(err);
	})
```

### Expected Behavior

Calling reply.callNotFound should work in this case.

### PR DESCRIPTION
**Pull request by VigneshMurugan**
#### Checklist

- [ ] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Review by mcollina — CHANGES_REQUESTED** [review]
Good work! Can you please add a unit test?

**Comment by VigneshMurugan** [comment]
> Good work! Can you please add a unit test?

I have added the unit case

File: test/404s.test.js
Comment (mcollina) [inline test/404s.test.js]: please revert all those changes

File: test/404s.test.js
Comment (mcollina) [inline test/404s.test.js]: I don't think this is needed

File: test/404s.test.js
Comment (VigneshMurugan) [reply test/404s.test.js]: Actually removed in latest commit

File: test/404s.test.js
Comment (Eomm) [inline test/404s.test.js:1776]:
```suggestion
test('Should handle to invoke callNotFound inside a 404 handler', t => {
```

File: test/404s.test.js
Comment (Eomm) [inline test/404s.test.js:1783]: could you add a test adding a not found handler?
https://github.com/fastify/fastify/blob/main/docs/Reference/Server.md#setnotfoundhandler

I would expect it is executed

File: test/404s.test.js
Comment (Eomm) [inline test/404s.test.js:1795]: What does the response body contain?

File: test/404s.test.js
Comment (mcollina) [inline test/404s.test.js:1788]: Please remove the try/catch, it's not needed for this test

File: test/404s.test.js
Comment (mcollina) [inline test/404s.test.js:1782]: please add an assertion inside this function

File: test/404s.test.js
Comment (VigneshMurugan) [reply test/404s.test.js:1783]: I thought it is handled in `Should fail to invoke callNotFound inside a 404 handler` test case. Should i write another testcase ?

File: test/404s.test.js
Comment (VigneshMurugan) [reply test/404s.test.js:1788]: Done

File: test/404s.test.js
Comment (Eomm) [reply test/404s.test.js:1783]: Yes please. That case does not contain a custom not found handler but it relies on the default one

File: test/404s.test.js
Comment (VigneshMurugan) [reply test/404s.test.js:1804]: done

File: test/404s.test.js
Comment (VigneshMurugan) [reply test/404s.test.js]: checked that as well

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by VigneshMurugan** [comment]
I don't have merge rights. So once @Eomm approves it can one of the maintainers merge it. @mcollina @jsumners

File: test/404s.test.js
Comment (Eomm) [inline test/404s.test.js:1812]: I think this assertion is wrong: I would expect to call this function from the `frameworkErrors`

It should be doable by updating the `onBadUrlContext` on:

```
function setNotFoundHandler (opts, handler) {
```

Comment (VigneshMurugan) [reply test/404s.test.js:1812]: I am sorry I didn't get your comment here. What I understand is that we should call `fastify.setNotFoundHandler` from `frameworkErrors` is it ?

setNotFoundHandler would be accessible post instantiation right ? but frameworkErrors have to be set during instantiation. 

I am bit confused here. Can you shed some light on it ?

Comment (Eomm) [reply test/404s.test.js:1812]: No, what I mean is that `reply.callNotFound()` "Invokes the custom not found handler" as written in the docs.

In this test, the `reply.callNotFound()` the custom not found handler is ignored by this fix.

I think it is correct to set the `onBadUrlContext[kFourOhFourContext]` to `null`, but the same field can be updated by editing this function:

https://github.com/fastify/fastify/blob/b15c7c566888c7953715c031a8b3bfe13c3d1960/fastify.js#L623

adding a `onBadUrlContext[kFourOhFourContext] = something`

Comment (VigneshMurugan) [reply test/404s.test.js:1812]: What I understood is that `reply.callNotFound()` invokes the preHandler `Invokes the custom not found handler. Note that it will only call 'preHandler' hook specified in ['setNotFoundHandler']` . But unfortunately in our case once we see the `reply.context[kFourOhFourContext] === null` we reply with 404 and the preHandler would never be called.

Comment (Eomm) [reply test/404s.test.js:1812]: I mean this:

```diff
diff --git a/fastify.js b/fastify.js
index 4a53945b..2507e21a 100644
--- a/fastify.js
+++ b/fastify.js
@@ -610,6 +610,7 @@ function fastify (options) {
 
       childLogger.info({ req }, 'incoming request')
 
+      onBadUrlContext[kFourOhFourContext] = this[kFourOhFourContext]
       const request = new Request(id, null, req, null, childLogger, onBadUrlContext)
       const reply = new Reply(res, request, childLogger)
       return frameworkErrors(new FST_ERR_BAD_URL(path), request, reply)
diff --git a/test/404s.test.js b/test/404s.test.js
index 5b7f13a1..f02764a0 100644
--- a/test/404s.test.js
+++ b/test/404s.test.js
@@ -1809,6 +1809,7 @@ test('Send 404 when frameworkError calls reply.callNotFound even with setNotFoun
     })
 
     fastify.setNotFoundHandler(function (_req, reply) {
+      t.pass('the custom 404 handler is called')
       reply.code(404).send('This should not be sent')
     })
 
@@ -1819,7 +1820,7 @@ test('Send 404 when frameworkError calls reply.callNotFound even with setNotFoun
     }, (err, response) => {
       t.error(err)
       t.equal(response.statusCode, 404)
-      t.equal(response.payload, '404 Not Found')
+      t.equal(response.payload, 'This should not be sent')
     })
   })
 

```

**Comment by mcollina** [comment]
@VigneshMurugan any update on this one?

**Comment by VigneshMurugan** [comment]
> @VigneshMurugan any update on this one?

Will update it tomorrow as i am still understanding the review comments by @Eomm

**Comment by VigneshMurugan** [comment]
@Eomm Let me explain my analysis

https://github.com/VigneshMurugan/fastify/blob/ccbbc780e90237ac7e1e4d332f63a9b07306ff13/fastify.js#L614 - This is where we call the frameworkErrors.

In https://github.com/VigneshMurugan/fastify/blob/ccbbc780e90237ac7e1e4d332f63a9b07306ff13/lib/reply.js#L701 We already return from the function. Hence preHandler of the custom not found handler wont be called. **The document clearly says that only the preHandler of the callNotFoundhandler would be called**

### .callNotFound()
<a id="call-not-found"></a>

Invokes the custom not found handler. Note that it will only call `preHandler`
hook specified in [`setNotFoundHandler`](./Server.md#set-not-found-handler).


Correct me if I am wrong @mcollina.

**Comment by Eomm** [comment]
As you wrote

> Invokes the custom not found handler

The custom "not found handler" is the function set by the user, it is not the default one.

In the new tests, the function set by the user is not called.

Anyway I'm not blocking this, this pr is already solving a bug.

**Comment by VigneshMurugan** [comment]
@mcollina I think @Eomm is ok with PR getting merged. Can you please merge it in case you if this LGT to you

**Pull request merged**
